sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java

```diff
@@ -132,37 +133,68 @@ public Node visitDropUser(SqlBaseParser.DropUserContext context) {
 
     @Override
     public Node visitGrantPrivilege(SqlBaseParser.GrantPrivilegeContext context) {
-        List<String> usernames = identsToStrings(context.ident());
+        List<String> usernames = identsToStrings(context.userNames().ident());
 
         if (context.ALL() != null) {
-            return new GrantPrivilege(usernames);
+            if (context.ON() == null) {
+                return new GrantPrivilege(usernames, "CLUSTER", Collections.EMPTY_LIST);
+            } else {
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
+                return new GrantPrivilege(usernames, getclazz(context.clazz().getStart()), tableOrSchemaNames);
+            }
         } else {
             List<String> privilegeTypes = identsToStrings(context.privilegeTypes().ident());
-            return new GrantPrivilege(usernames, privilegeTypes);
+            if (context.ON() == null) {
+                return new GrantPrivilege(usernames, privilegeTypes, "CLUSTER", Collections.EMPTY_LIST);
+            } else {
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
+
+                return new GrantPrivilege(usernames, privilegeTypes, getclazz(context.clazz().getStart()), tableOrSchemaNames);
+            }
         }
     }
 
     @Override
     public Node visitDenyPrivilege(SqlBaseParser.DenyPrivilegeContext context) {
-        List<String> usernames = identsToStrings(context.ident());
+        List<String> usernames = identsToStrings(context.userNames().ident());
 
         if (context.ALL() != null) {
-            return new DenyPrivilege(usernames);
+            if (context.ON() == null) {
+                return new DenyPrivilege(usernames, "CLUSTER", Collections.EMPTY_LIST);
+            } else {
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
+                return new DenyPrivilege(usernames, getclazz(context.clazz().getStart()), tableOrSchemaNames);
+            }
         } else {
             List<String> privilegeTypes = identsToStrings(context.privilegeTypes().ident());
-            return new DenyPrivilege(usernames, privilegeTypes);
+            if (context.ON() == null) {
+                return new DenyPrivilege(usernames, privilegeTypes, "CLUSTER", Collections.EMPTY_LIST);
+            } else {
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
```

Comment:
would be great to factor this out into a method e.g. `getClassAndIdentsForPrivileges(...)` which can then be used in all 3 privilege node methods